chainstate/src/detail/transaction_verifier/mod.rs

```diff
@@ -489,6 +577,212 @@ impl<'a, S: BlockchainStorageRead> TransactionVerifier<'a, S> {
+                    TokenData::TokenTransferV1 { token_id, amount } => {
+                        (token_id == burn_token_id).then_some(amount)
+                    }
+                    TokenData::TokenIssuanceV1 {
+                        token_ticker: _,
+                        amount_to_issue: _,
+                        number_of_decimals: _,
+                        metadata_uri: _,
+                    }
+                    | TokenData::TokenBurnV1 {
+                        token_id: _,
+                        amount_to_burn: _,
+                    } => None,
+                },
+            })
+            .try_fold(Amount::from_atoms(0), |accum, output| accum + *output)
+            .ok_or(TokensError::CoinOrTokenOverflow)?;
+        Ok(change_amount)
+    }
+
+    fn register_issuance(&mut self, tx: &Transaction) -> Result<(), TokensError> {
+        let token_id = token_id(tx).ok_or(TokensError::TokenIdCantBeCalculated)?;
+
+        let _tokens_op = match self.tokens_cache.entry(token_id) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(CachedTokensOperation::Write(tx.get_id())),
+        };
+        Ok(())
+    }
+
+    fn undo_issuance(&mut self, token_id: TokenId) -> Result<(), TokensError> {
+        match self.tokens_cache.entry(token_id) {
+            Entry::Occupied(entry) => {
+                let tokens_op = entry.into_mut();
+                *tokens_op = CachedTokensOperation::Erase;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(CachedTokensOperation::Erase);
+            }
+        }
+        Ok(())
+    }
+
+    fn check_connected_tokens(
+        &mut self,
+        token_data: &TokenData,
+        tx: &Transaction,
+        block: &Block,
+    ) -> Result<(), TokensError> {
+        match token_data {
+            TokenData::TokenTransferV1 { token_id, amount } => {
+                self.check_connected_transfer_data(block.get_id(), tx, token_id, amount)?;
+            }
+            TokenData::TokenIssuanceV1 {
+                token_ticker,
+                amount_to_issue,
+                number_of_decimals,
+                metadata_uri,
+            } => {
+                self.check_connected_issue_data(
+                    token_ticker,
+                    amount_to_issue,
+                    number_of_decimals,
+                    metadata_uri,
+                    tx.get_id(),
+                    block.get_id(),
+                )?;
+                self.register_issuance(tx)?;
+            }
+            TokenData::TokenBurnV1 {
+                token_id,
+                amount_to_burn,
+            } => {
+                self.check_connected_burn_data(tx, &block.get_id(), token_id, amount_to_burn)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn check_tokens_values(&mut self, block: &Block) -> Result<(), ConnectTransactionError> {
```

Comment:
Please move this and all other token-related stuff to a submodule. No need to make this a bigger pan of soup of unrelated things. The file is already 1000 lines long.

Reply:
Also, "check" can never be mutable.